src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,346 @@
+                  iterable_right.with_iterators([&](auto right_it, auto right_end) {
+                      with_comparator(_predicate_condition, [&](auto comparator) {
+                          this->_join_two_columns(comparator, left_it, left_end, right_it, right_end, chunk_id_left,
+                                                  chunk_id_right, left_matches, index);
+                      });
+                  });
+              });
+          }
+          // clang-format on
+        });
+      });
+    }
+  }
+
+  // For Full Outer and Left Join we need to add all unmatched rows for the left side
+  if (_mode == JoinMode::Left || _mode == JoinMode::Outer) {
+    for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < _left_in_table->chunk_count(); ++chunk_id_left) {
+      for (ChunkOffset chunk_offset{0}; chunk_offset < _left_matches[chunk_id_left].size(); ++chunk_offset) {
+        if (!_left_matches[chunk_id_left][chunk_offset]) {
+          _pos_list_left->emplace_back(RowID{chunk_id_left, chunk_offset});
+          _pos_list_right->emplace_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+        }
+      }
+    }
+  }
+
+  // For Full Outer and Right Join we need to add all unmatched rows for the right side.
+  if (_mode == JoinMode::Outer || _mode == JoinMode::Right) {
+    for (ChunkID chunk_id{0}; chunk_id < _right_matches.size(); ++chunk_id) {
+      for (ChunkOffset chunk_offset{0}; chunk_offset < _right_matches[chunk_id].size(); ++chunk_offset) {
+        if (!_right_matches[chunk_id][chunk_offset]) {
+          _pos_list_right->emplace_back(RowID{chunk_id, chunk_offset});
+          _pos_list_left->emplace_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+        }
+      }
+    }
+  }
+
+  // write output chunks
+  auto output_chunk = std::make_shared<Chunk>();
+
+  _write_output_chunk(output_chunk, _left_in_table, _pos_list_left);
+  _write_output_chunk(output_chunk, _right_in_table, _pos_list_right);
+
+  _output_table->emplace_chunk(output_chunk);
+}
+
+// join loop that joins two chunks of two columns via their iterators
+template <typename BinaryFunctor, typename LeftIterator, typename RightIterator>
+void JoinIndex::_join_two_columns(const BinaryFunctor& func, LeftIterator left_it, LeftIterator left_end,
+                                  RightIterator right_begin, RightIterator right_end, const ChunkID chunk_id_left,
+                                  const ChunkID chunk_id_right, std::vector<bool>& left_matches,
+                                  std::shared_ptr<BaseIndex> index) {
+  if (index != nullptr) {
+    // We have a proper index and can use it, so we do an index join for this chunk combination
+
+    for (; left_it != left_end; ++left_it) {
+      const auto left_value = *left_it;
+      if (left_value.is_null()) continue;
+
+      // The index interface takes a vector of ATVs, so we need to wrap our value
+      auto comp_values = std::vector<AllTypeVariant>{AllTypeVariant(left_value.value())};
```

Comment:
Do you know how your join compares to the other joins? You are doing one or two virtual method call per left column value at the moment.

Reply:
This virtual call is due to the fact that we can not get an all type variant from the iterators but one is required by the index interface. If there was a type erased iterator we could get on without this call.
Apart from that we compare favourably to all current iterators when the right column is sufficiently large. The virtual calls have not shown up as hotspots in our profiling runs so we believe them to be uncritical. Still makes me unhappy though.